src/native/minipal/utf8.h

```diff
@@ -0,0 +1,82 @@
+ * @param sourceLength Length of the source string.
+ * @param destination Pointer to the destination UTF-16 string. It can be NULL to query number of items required by the conversion.
+ * @param destinationLength Length of the destination string.
+ * @param flags Flags to alter the behavior of converter. Supported flags are MINIPAL_MB_ERR_INVALID_CHARS and MINIPAL_TREAT_AS_LITTLE_ENDIAN.
+ * @return Number of items required by the conversion.
+ */
+size_t minipal_utf8_to_utf16_preallocated(const char* source, size_t sourceLength, CHAR16_T** destination, size_t destinationLength, unsigned int flags);
+
+/**
+ * Convert a string from UTF-16 (UCS-2) to UTF-8 with preallocated memory
+ *
+ * @param source The source string in UTF-16 format.
+ * @param sourceLength Length of the source string.
+ * @param destination Pointer to the destination UTF-8 string. It can be NULL to query number of items required by the conversion.
+ * @param destinationLength Length of the destination string.
+ * @param flags Flags to alter the behavior of converter. Supported flags are MINIPAL_MB_ERR_INVALID_CHARS and MINIPAL_TREAT_AS_LITTLE_ENDIAN.
+ * @return Number of items required by the conversion.
+ */
+size_t minipal_utf16_to_utf8_preallocated(const CHAR16_T* source, size_t sourceLength, char** destination, size_t destinationLength, unsigned int flags);
+
+/**
+ * Convert a string from UTF-16 (UCS-2) to UTF-8 by allocating memory
+ *
+ * The caller of this API should call free() on the `destination` pointer after the usage.
+ *
+ * @param source The source string in UTF-16 format.
+ * @param sourceLength Length of the source string.
+ * @param destination Pointer to the destination UTF-8 string. It can be NULL to query number of items required by the conversion.
+ * @param destinationLength Length of the destination string.
+ * @param flags Flags to alter the behavior of converter. Supported flags are MINIPAL_MB_ERR_INVALID_CHARS and MINIPAL_TREAT_AS_LITTLE_ENDIAN.
+ * @return Number of items required by the conversion.
+ */
+size_t minipal_utf8_to_utf16_allocate(const char* source, size_t sourceLength, CHAR16_T** destination, unsigned int flags);
+
+/**
+ * Convert a string from UTF-16 (UCS-2) to UTF-8 with preallocated memory
+ *
+ * The caller of this API should call free() on the `destination` pointer after the usage.
+ *
+ * @param source The source string in UTF-16 format.
+ * @param sourceLength Length of the source string.
+ * @param destination Pointer to the destination UTF-8 string. It can be NULL to query number of items required by the conversion.
+ * @param destinationLength Length of the destination string.
+ * @param flags Flags to alter the behavior of converter. Supported flags are MINIPAL_MB_ERR_INVALID_CHARS and MINIPAL_TREAT_AS_LITTLE_ENDIAN.
+ * @return Number of items required by the conversion.
+ */
+size_t minipal_utf16_to_utf8_allocate(const CHAR16_T* source, size_t sourceLength, char** destination, unsigned int flags);
```

Comment:
These `_allocate` methods are used in exactly one place in Mono. Do we really need them?

(They are a bit weird in that they take buffer and return zero terminated string. If these were meant to be convenience helpers, I would expect them to take zero terminated string and return zero terminated string.)

Reply:
The input string is also zero-terminated. Otherwise, strlen/wcslen on input fail to produce the correct count that show up in CI testing.

Reply:
If the input is zero terminated, the method should not take the `sourceLength` argument.

Reply:
When `sourceLength` > 0, nulls are ignored (`string with nulls` case). We need both.

Reply:
Can it be handled by the caller instead?

Reply:
Yes, it can be done differently. Current set of APIs are either allocating or use preallocated memory. I was thinking that we will remove the coreclr/mono helpers in the future, and make the final callers to consume these APIs directly. Which is why I've favored PAL/eglib intermediate callers to be relatively slimmer with these convenient APIs.

Reply:
I do like the idea of deleting unnecessary layers. At the same time, I think it is important to have simple clean contracts for the minipal APIs. Changing the API behavior when it is based buffer with zero length does not sound like a simple clean contract.

Reply:
utf8.cpp was (partially) following win32 API conventions, now we have converted try-catch with asserts and errno in utf8.c version, it makes sense to rethink the shape suitable for minipal.

Overall, consumers need four things from minipal:

* query number of chars required post-conversion for char and char16_t
* convert char to char16_t and vice-versa

We can stop chasing win32 API convention and provide four separate APIs for them, and let the callers handle the rest:

```h
// will assert(sourceLength > 0)
size_t minipal_get_length_utf8_to_utf16(const char* source, size_t sourceLength, unsigned int flags);

// will assert(sourceLength > 0)
size_t minipal_get_length_utf16_to_utf8(const CHAR16_T* source, size_t sourceLength, unsigned int flags);

// will assert(sourceLength > 0 && destinationLength > 0)
// will NOT set destination[destinationLength - 1] = '\0'
void minipal_convert_utf8_to_utf16(const char* source, size_t sourceLength, CHAR16_T** destination, size_t destinationLength, unsigned int flags);

// will assert(sourceLength > 0 && destinationLength > 0)
// will NOT set destination[destinationLength - 1] = '\0'
void minipal_convert_utf16_to_utf8(const CHAR16_T* source, size_t sourceLength, char** destination, size_t destinationLength, unsigned int flags);
```

Does this look reasonable?

Reply:
> will assert(sourceLength > 0)
> will assert(sourceLength > 0 && destinationLength > 0)

Can these APIs deal with the zero-sized buffers gracefully? For example, `Encoding.GetBytes(ReadOnlySpan<Char>, Span<Byte>)`  managed API deals with zero-sized buffers gracefully as well.

>  minipal_convert_

Should the convert methods return how much was written into the destination buffer?

LGTM otherwise.

Reply:
> CHAR16_T** destination
> char** destination

I assume that these should be `CHAR16_T*` and `char*`.